spanner_orm/model.py

```diff
@@ -327,11 +241,94 @@ def _execute_write(cls, db_api, transaction, dictionaries):
             'Attempted to update rows with different sets of keys')
 
       for key, value in dictionary.items():
-        cls._validate(key, value, error.SpannerError)
+        cls.validate_value(key, value, error.SpannerError)
       values.append([dictionary[column] for column in columns])
 
-    args = [cls.table(), columns, values]
+    args = [cls.table, columns, values]
     if transaction is not None:
       return db_api(transaction, *args)
     else:
       return api.SpannerApi.run_write(db_api, *args)
+
+
+class Model(object, metaclass=ModelMeta):
+  """Maps to a table in spanner and has basic functions for querying tables."""
+
+  # Instance methods
+  def __init__(self, values, persisted=False):
+    metaclass = type(self)
+    self.related = {}
+    self.start_values = {}
+    # Ensure that we have the primary index keys (unique id) set for all objects
+    missing_keys = set(metaclass.primary_keys) - set(values.keys())
+    if missing_keys:
+      raise error.SpannerError(
+          'All primary keys must be specified. Missing: {keys}'.format(
+              keys=missing_keys))
+
+    for key, value in values.items():
+      if key in metaclass.relations:
+        self.related[key] = value
+      elif key in metaclass.columns:
+        if value is not None:
+          metaclass.validate_value(key, value, ValueError)
+          self.start_values[key] = copy.copy(value)
+      else:
+        raise ValueError('{name} is not part of {klass}'.format(
+            name=key, klass=metaclass.__name__))
+
+    self.values = copy.deepcopy(self.start_values)
+    self._persisted = persisted
+
+  def __getattr__(self, name):
+    metaclass = type(self)
+    if name in metaclass.schema:
+      return self.values.get(name)
+    elif name in metaclass.relations:
+      if name in self.related:
+        return self.related[name]
+      raise AttributeError('{name} was not included in query'.format(name=name))
+    raise AttributeError(name)
+
+  def __setattr__(self, name, value):
+    metaclass = type(self)
+    if name in metaclass.primary_keys:
+      raise AttributeError(name)
+    elif name in metaclass.schema:
+      metaclass.validate_value(name, value, AttributeError)
+      self.values[name] = copy.copy(value)
+    elif name in metaclass.relations:
+      raise AttributeError('{name} is not settable'.format(name=name))
+    else:
+      super().__setattr__(name, value)
+
+  def changes(self):
+    metaclass = type(self)
+    return {
+        key: self.values[key]
+        for key in metaclass.columns
+        if self.values.get(key) != self.start_values.get(key)
+    }
+
+  def id(self):
+    metaclass = type(self)
+    return {key: self.values[key] for key in metaclass.primary_keys}
```

Comment:
I can imagine people wanting to get the metaclass info from an instance.  Can we wrap each metaclass call as a function:
```
def _metaclass(self):
  return type(self)

def primary_keys(self):
  self._metaclass.primary_keys()

def columns(self):
  self._metaclass.columns()

...
```
Also, if we have each of these in a function we change how we store one of these, then we only have to change it in one place as opposed to several.

Reply:
i did this intentionally because i couldn't think of a reason people would need to get metaclass info from an instance. do you have an example of where someone would need this?

Reply:
If someone wants a function that can take a few types of models as an input (like navy with resources), they might want to be able to access columns:
```
expose_to_client = {'name', 'expiration_date', ...}
def exposable_fields_from_resource(resource):
  return [resource[field_name] for field_name in set(resource.columns()).union(expose_to_client)]
```
I don't think it's a huge requirement, but I do think it's a nice convenience.  I also think it moves a lot of similar logic in `Model` code to a couple places. 

Are there any reasons you think we should explicitly not reveal these?

Reply:
interesting, are you talking about the definition of classes that derive from Model, or are you talking about people who hold an instance of a Model object? if it's the former, i can see your point. i guess i'd be ok with adding them as private properties

Reply:
Yea the former.  So yea that works